### PR TITLE
8343286: Missing unchecked cast warning in polymorphic method call

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
Javac is not issuing a mandated unchecked cast warning for code like:

```
import java.lang.invoke.VarHandle;

class VarHandleCast<V> {
     VarHandle vh;
     V method(Object obj) {
         return (V)vh.getAndSet(this, obj);
     }
}
```

according to the spec the return type inferred for this method polymorphic method should be Object see:
```
-  If the compile-time declaration for the method invocation is a signature polymorphic method, then:
...
    – The compile-time result is determined as follows:
        Otherwise, if the method invocation expression is the operand of a cast
    expression (§15.16), the compile-time result is the erasure of the type of the
    cast expression (§4.6).
```
javac is not doing the type erasure mandated by the spec which is what this PR is fixing

TIA

Note: I had to add some annotations to suppress warnings in java.base that were not being issued before due to the javac bug

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8343733](https://bugs.openjdk.org/browse/JDK-8343733) to be approved

### Issues
 * [JDK-8343286](https://bugs.openjdk.org/browse/JDK-8343286): Missing unchecked cast warning in polymorphic method call (**Bug** - P4)
 * [JDK-8343733](https://bugs.openjdk.org/browse/JDK-8343733): Missing unchecked cast warning in polymorphic method call (**CSR**)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21914/head:pull/21914` \
`$ git checkout pull/21914`

Update a local copy of the PR: \
`$ git checkout pull/21914` \
`$ git pull https://git.openjdk.org/jdk.git pull/21914/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21914`

View PR using the GUI difftool: \
`$ git pr show -t 21914`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21914.diff">https://git.openjdk.org/jdk/pull/21914.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21914#issuecomment-2458039471)
</details>

### CODE REVIEW

**Comment by jddarcy** [comment]
Hi @vicente-romero-oracle , please file a CSR to evaluate the behavioral changes here. Thanks.

**Review by mcimadamore — APPROVED** [review]
Changes look good. The extra suppress warnings seem to indicate that the logic is now working as intended :-)

**Comment by vicente-romero-oracle** [comment]
> Changes look good. The extra suppress warnings seem to indicate that the logic is now working as intended :-)

thanks for the review. As suggested above I created a CSR, could you please review it too? Thanks!

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Source.java [hunk 1]
```diff
@@ -263,6 +263,7 @@ public enum Feature {
         FLEXIBLE_CONSTRUCTORS(JDK22, Fragments.FeatureFlexibleConstructors, DiagKind.NORMAL),
         MODULE_IMPORTS(JDK23, Fragments.FeatureModuleImports, DiagKind.PLURAL),
         PRIVATE_MEMBERS_IN_PERMITS_CLAUSE(JDK19),
+        COMPILE_TIME_TYPE_SIG_POLYMORPHIC_MTH(JDK24),
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Source.java:266]: Nit: I'd pick a name that is more representative of what the behavior is. E.g. `erase_poly_sig_return_type`. Same for the field in `Infer` storing the boolean flag.

**Review by mcimadamore — APPROVED** [review]

**Pull request closed without merging**